gadget/install/install.go

```diff
@@ -377,6 +378,160 @@ func Run(model gadget.Model, gadgetRoot, kernelRoot, bootDevice string, options
+			}
+		}
+	}
+
+	return nil, fmt.Errorf("cannot find laid out structure for %q", onDiskStruct.Name)
+}
+
+// sysfsPathForBlockDevice returns the sysfs path for a block device.
+func sysfsPathForBlockDevice(device string) (string, error) {
+	partPath, err := os.Readlink(filepath.Join("/sys/class/block", device))
+	if err != nil {
+		return "", err
+	}
+	// Remove initial ../../ and make path absolute
+	return filepath.Join("/sys", partPath[6:]), nil
+}
+
+// onDiskVolumeFromPartitionSysfsPath creates an OnDiskVolume that
+// matches the disk that contains the given partition sysfs path
+func onDiskVolumeFromPartitionSysfsPath(partPath string) (*gadget.OnDiskVolume, error) {
+	// Removing the last component will give us the disk path
+	diskPath := filepath.Dir(partPath)
+	disk, err := disks.DiskFromDevicePath(diskPath)
+	if err != nil {
+		return nil, fmt.Errorf("cannot retrieve disk information for %q: %v", partPath, err)
+	}
+	onDiskVol, err := gadget.OnDiskVolumeFromDisk(disk)
+	if err != nil {
+		return nil, fmt.Errorf("cannot retrieve on disk volume for %q: %v", partPath, err)
+	}
+
+	return onDiskVol, nil
+}
+
+// buildOnDiskStruct obtains the current on disk structure from a
+// sysfs partition path string. Also, we take the laid out information
+// from laidOutVols and insert it in the disk structure.
+func buildOnDiskStruct(onDiskVol *gadget.OnDiskVolume, partPath string, laidOutVols map[string]*gadget.LaidOutVolume, gadgetVolName string) (*gadget.OnDiskStructure, error) {
+	partName := filepath.Base(partPath)
+	partIdx, err := partIndexFromPartName(partName)
+	if err != nil {
+		return nil, fmt.Errorf("cannot find partition index in %q: %v", partPath, err)
+	}
+	onDiskStruct, err := structureFromIndex(onDiskVol, partIdx)
+	if err != nil {
+		return nil, fmt.Errorf("cannot find partition %q: %v", partPath, err)
+	}
+
+	laidOutStruct, err := laidOutStructForDiskStruct(laidOutVols, gadgetVolName, onDiskStruct)
+	if err != nil {
+		return nil, err
+	}
+	// Fill ResolvedContent
+	onDiskStruct.LaidOutStructure = *laidOutStruct
+
+	return onDiskStruct, nil
+}
+
+// WriteContent writes gadget content to the devices specified in
+// onVolumes. It returns the resolved on disk volumes.
+func WriteContent(onVolumes map[string]*gadget.Volume, observer gadget.ContentObserver,
+	gadgetRoot, kernelRoot string, model *asserts.Model, perfTimings timings.Measurer) ([]*gadget.OnDiskVolume, error) {
+
+	_, allLaidOutVols, err := gadget.LaidOutVolumesFromGadget(gadgetRoot, kernelRoot, model)
+	if err != nil {
+		return nil, fmt.Errorf("when writing content: cannot layout volumes: %v", err)
+	}
+
+	var onDiskVols []*gadget.OnDiskVolume
+	for volName, vol := range onVolumes {
+		var onDiskVol *gadget.OnDiskVolume
+		for _, volStruct := range vol.Structure {
+			// TODO fixme
+			if volStruct.Role == "mbr" {
+				continue
+			}
+			// TODO write raw content?
+			if volStruct.Role == "" {
+				continue
+			}
+
+			device := filepath.Base(volStruct.Device)
```

Comment:
should we validate somehow .Device before processing it?

Reply:
Somewhat we already validate it, first when we call Readlink (it will fail if not a block device), then when we try to find the disk, etc. Not sure if adding a test previous to that would make much sense.